source/Mlos.Model.Services.Client/OptimizerServiceEncoderDecoder.cs

```
@@ -0,0 +1,419 @@
+            foreach (var objective in problem.Objectives)
+            {
+                instance.Objectives.Add(EncodeObjective(objective));
+            }
+
+            if (problem.ContextSpace == null)
+            {
+                instance.EmptyContext = new OptimizerService.Empty();
+            }
+            else
+            {
+                instance.ContextSpace = EncodeHypergrid(problem.ContextSpace);
+            }
+
+            return instance;
+        }
+
+        public OptimizerService.Objective EncodeObjective(OptimizationObjective objective)
+        {
+            return new OptimizerService.Objective
+            {
+                Minimize = objective.Minimize,
+                Name = objective.Name,
+            };
+        }
+
+        public OptimizerService.GuestSubgrid EncodeSubgridJoin(SubgridJoin subgridJoin)
+        {
+            return new OptimizerService.GuestSubgrid
+            {
+                ExternalPivotDimension = EncodeDimension(subgridJoin.OnExternalJoin),
+                Subgrid = EncodeHypergrid(subgridJoin.Subgrid),
+            };
+        }
+
+        private OptimizerService.DimensionType DimensionToGrpcType(DimensionTypeName typeName)
+        {
+            switch (typeName)
+            {
+                case DimensionTypeName.CategoricalDimension:
+                    return OptimizerService.DimensionType.Categorical;
+                case DimensionTypeName.ContinuousDimension:
+                    return OptimizerService.DimensionType.Continuous;
+                case DimensionTypeName.DiscreteDimension:
+                    return OptimizerService.DimensionType.Discrete;
+                case DimensionTypeName.OrdinalDimension:
+                    return OptimizerService.DimensionType.Ordinal;
+                default:
+                    System.Console.Error.WriteLine("Invalid type provided to OptimizerServiceEncoderDecoder.Encoder.DimensionToGrpcType(..). Type: " + typeName);

```

We should just throw an exception. Assuming a continuous dimension was intended has something like 25% of being correct. Better to break things early.

---

Also, why not use a dictionary to map types between the two different type systems?

---

> We should just throw an exception. Assuming a continuous dimension was intended has something like 25% of being correct. Better to break things early.

I was coming from my Java days where throwing exceptions makes code a total pain to deal with. Adding now. 

> dictionary

I couldn't figure out how to do an inline initializer for a C# dictionary. Writing a switch statement was faster. 
Can you do inline initializers for C# dictionaries?